### PR TITLE
extend ConditionalFact and ConditionalTheory to support dynamic skip

### PR DESCRIPTION
**Pull request by wfurt**
We can already implement custom function to skip tests in specific cases. That however must be static method and that has some limitations. HTTP in corefx for example uses tests where interesting variable like UseSocketsHttpHandler are set on instance derived from base tests. With this change, we can do something like:

```c#
-            if (IsCurlHandler) return;
+            if (IsCurlHandler) {
+                throw new SkipTestException("Not supported with CurlHandler");
+            }
```

And when executed it can produce  something like:
```
System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpCookieProtocolTests.GetAsyncWithRedirect_SetCookieContainer_CorrectCookiesSent [STARTING]
System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpCookieProtocolTests.GetAsyncWithRedirect_SetCookieContainer_CorrectCookiesSent [FINISHED] Time: 0.1237862s
System.Net.Http.Functional.Tests.PlatformHandler_HttpCookieProtocolTests.GetAsyncWithRedirect_SetCookieContainer_CorrectCookiesSent [STARTING]
System.Net.Http.Functional.Tests.PlatformHandler_HttpCookieProtocolTests.GetAsyncWithRedirect_SetCookieContainer_CorrectCookiesSent [SKIP]
System.Net.Http.Functional.Tests.PlatformHandler_HttpCookieProtocolTests.GetAsyncWithRedirect_SetCookieContainer_CorrectCookiesSent [FINISHED] Time: 0.0003453s
```

Also testResult.xml would have expected output as 
```
      <test name="System.Net.Http.Functional.Tests.PlatformHandler_HttpCookieProtocolTests.GetAsyncWithRedirect_SetCookieContainer_CorrectCookiesSent" type="System.Net.Http.Functional.Tests.PlatformHandler_HttpCookieProtocolTests" method="GetAsyncWithRedirect_SetCookieContainer_CorrectCookiesSent" time="0" result="Skip">
      <test name="System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpCookieProtocolTests.GetAsyncWithRedirect_SetCookieContainer_CorrectCookiesSent" type="System.Net.Http.Functional.Tests.SocketsHttpHandler_HttpCookieProtocolTests" method="GetAsyncWithRedirect_SetCookieContainer_CorrectCookiesSent" time="0.1237862" result="Pass" />

``` 

and we could properly track skipped tests instead of reporting lame pass.

### CODE REVIEW

File: src/Microsoft.DotNet.XUnitExtensions/src/SkippedTestCase.cs [hunk 1]
```diff
@@ -12,6 +13,102 @@
...
+
+            // Nothing we care about, send it on its way
+            return innerBus.QueueMessage(message);
+        }
+    }
+
+    /// <summary>Wraps RunAsync for ConditionalTheory.</summary>
+    public class SkippedTheoryTestCase : XunitTheoryTestCase
+    {
+        [Obsolete("Called by the de-serializer; should only be called by deriving classes for de-serialization purposes")]
```
Comment (sharwell) [inline src/Microsoft.DotNet.XUnitExtensions/src/SkippedTestCase.cs]: 💡 

```suggestion
        [Obsolete("Called by the de-serializer; should only be called by deriving classes for de-serialization purposes", error: true)]
```

File: src/Microsoft.DotNet.XUnitExtensions/src/SkippedTestCase.cs [hunk 2]
```diff
@@ -12,6 +13,102 @@
...
+            }
+
+            return result;
+        }
+    }
+
+    /// <summary>Wraps RunAsync for ConditionalFact.</summary>
+    public class SkippedFactTestCase : XunitTestCase
+    {
+        [Obsolete("Called by the de-serializer; should only be called by deriving classes for de-serialization purposes")]
```
Comment (sharwell) [inline src/Microsoft.DotNet.XUnitExtensions/src/SkippedTestCase.cs]: 💡 

```suggestion
        [Obsolete("Called by the de-serializer; should only be called by deriving classes for de-serialization purposes", error: true)]
```

File: src/Microsoft.DotNet.XUnitExtensions/src/SkippedTestCase.cs [hunk 3]
```diff
@@ -12,6 +13,102 @@
...
+                }
+            }
+
+            // Nothing we care about, send it on its way
+            return innerBus.QueueMessage(message);
+        }
+    }
+
+    /// <summary>Wraps RunAsync for ConditionalTheory.</summary>
+    public class SkippedTheoryTestCase : XunitTheoryTestCase
```
Comment (sharwell) [inline src/Microsoft.DotNet.XUnitExtensions/src/SkippedTestCase.cs]: 💡 I have a strong preference for one top-level type per file, but I'm not sure what this repository uses as a standard style.

**Comment by sharwell** [comment]
💡 Consider filing a bug for xUnit requesting an equivalent to `Assert.Inconclusive`.

File: src/Microsoft.DotNet.XUnitExtensions/src/SkippedTestCase.cs [hunk 4]
```diff
@@ -12,6 +13,102 @@
...
 {
+
+    public class SkipTestException : Exception
+    {
+        public SkipTestException(string reason)
+            : base(reason) { }
+    }
+
+    /// <summary>Implements message buss to communicate tests skipped via SkipTestException.</summary>
+    public class ConditionalTestMessageBus : IMessageBus
```
[the remark this answers is not in the document]
Comment (wfurt) [reply src/Microsoft.DotNet.XUnitExtensions/src/SkippedTestCase.cs]: done

**Pull request merged**